src/snapshot/src/lib.rs

```diff
@@ -0,0 +1,813 @@
+    where
+        T: Versionize,
+    {
+        let mut new_section = Section {
+            name: name.to_owned(),
+            data: vec![0; SNAPSHOT_MAX_SECTION_SIZE],
+        };
+
+        let slice = &mut new_section.data.as_mut_slice();
+        object.serialize(slice, &self.version_map, self.target_version)?;
+        // Resize vec to serialized section len.
+        let serialized_len =
+            slice.as_ptr() as usize - new_section.data.as_slice().as_ptr() as usize;
+        new_section.data.truncate(serialized_len);
+        self.sections.insert(name.to_owned(), new_section);
+        Ok(serialized_len)
+    }
+
+    // Returns the current snapshot format version.
+    // Not to be confused with data version which refers to the aplication
+    // defined structures.
+    // This version map allows us to change the underlying storage format -
+    // for example the way we encode vectors or moving to something else than bincode.
+    fn format_version_map() -> VersionMap {
+        // Firecracker snapshot format version 1.
+        VersionMap::new()
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    #![allow(non_upper_case_globals)]
+    #![allow(non_camel_case_types)]
+    #![allow(non_snake_case)]
+    use super::*;
+
+    static basic_add_remove_field_v1: &[u8] = &[
+        0x01, 0x00, 0x64, 0x86, 0x84, 0x19, 0x10, 0x07, 0x01, 0x00, 0x01, 0x00, 0x04, 0x00, 0x00,
+        0x00, 0x00, 0x00, 0x00, 0x00, 0x74, 0x65, 0x73, 0x74, 0x18, 0x00, 0x00, 0x00, 0x00, 0x00,
+        0x00, 0x00, 0xD2, 0x04, 0x00, 0x00, 0x04, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x74,
+        0x65, 0x73, 0x74, 0x14, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+    ];
+
+    static basic_add_remove_field_v2: &[u8] = &[
+        0x01, 0x00, 0x64, 0x86, 0x84, 0x19, 0x10, 0x07, 0x02, 0x00, 0x01, 0x00, 0x04, 0x00, 0x00,
+        0x00, 0x00, 0x00, 0x00, 0x00, 0x74, 0x65, 0x73, 0x74, 0x21, 0x00, 0x00, 0x00, 0x00, 0x00,
+        0x00, 0x00, 0x04, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x74, 0x65, 0x73, 0x74, 0x05,
+        0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x62, 0x61, 0x73, 0x69, 0x63, 0x14, 0x00, 0x00,
+        0x00, 0x00, 0x00, 0x00, 0x00,
+    ];
+
+    static basic_add_remove_field_v3: &[u8] = &[
+        0x01, 0x00, 0x64, 0x86, 0x84, 0x19, 0x10, 0x07, 0x03, 0x00, 0x01, 0x00, 0x04, 0x00, 0x00,
+        0x00, 0x00, 0x00, 0x00, 0x00, 0x74, 0x65, 0x73, 0x74, 0x29, 0x00, 0x00, 0x00, 0x00, 0x00,
+        0x00, 0x00, 0x04, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x74, 0x65, 0x73, 0x74, 0x05,
+        0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x62, 0x61, 0x73, 0x69, 0x63, 0xD2, 0x04, 0x00,
+        0x00, 0x00, 0x00, 0x00, 0x00, 0x14, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+    ];
+
+    #[derive(Versionize, Clone, Default, Debug)]
+    pub struct Test1 {
+        field_x: u64,
+        field0: u64,
+        field1: u32,
+    }
+
+    #[derive(Versionize, Clone, Default, Debug)]
+    pub struct Test {
```

Comment:
I think it would be useful to add `union`, `enum` and particularly `Option` fields as well to cover all sorts of weird transitions

Reply:
The plan is to cover all functionality for all the primitives as well as combinations, however we will do this in a future PR.